### PR TITLE
Hugging Face encoder-decoder architecture support

### PR DESCRIPTION
**Pull request by BlackSamorez**
T5 in particular

### CODE REVIEW

File: src/tensor_parallel/slicing_configs.py [hunk 1]
```diff
@@ -85,6 +85,46 @@ def split_alibi(alibi: torch.Tensor, *, rank: int, num_heads: int, world_size: i
...
+
+    select_kv_for_rank = lambda kvs, rank: (kvs[2 * rank], kvs[2 * rank + 1]) if kvs else None
+
+    return Config(
+        state_rules={
+            r".*SelfAttention\.q\.(weight|bias)": partial(split_heads, dim=0, head_dim=head_dim, world_size=world_size),
+            r".*SelfAttention\.k\.(weight|bias)": partial(split_heads, dim=0, head_dim=head_dim, world_size=world_size),
+            r".*SelfAttention\.v\.(weight|bias)": partial(split_heads, dim=0, head_dim=head_dim, world_size=world_size),
+            r".*relative_attention_bias\.weight": "split 1",
+            r".*SelfAttention\.o\.weight": "split 1",
```
Comment (justheuristic) [inline src/tensor_parallel/slicing_configs.py:104]: please check for $ at the end

File: src/tensor_parallel/slicing_configs.py [hunk 2]
```diff
@@ -85,6 +85,46 @@ def split_alibi(alibi: torch.Tensor, *, rank: int, num_heads: int, world_size: i
     return alibi_part.reshape(-1, *alibi.shape[1:])
 
 
+def get_t5_config(model_config: T5Config, devices: Sequence[torch.device]) -> Config:
+    world_size = len(devices)
+    num_heads = model_config.num_heads
+    head_dim = model_config.d_kv
+    gather_kv_across_ranks = CollectiveOperation(
```
Comment (justheuristic) [inline src/tensor_parallel/slicing_configs.py:92]: please make sure you use this op

File: setup.cfg [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [metadata]
 name = tensor_parallel
-version = 1.0.3
+version = 1.0.4dev
```
Comment (justheuristic) [inline setup.cfg:3]: 1.0.22?

Comment (BlackSamorez) [reply setup.cfg]: 1.0.30

File: src/tensor_parallel/slicing_configs.py (hunk 1 above)
Comment (BlackSamorez) [reply src/tensor_parallel/slicing_configs.py]: Done

File: src/tensor_parallel/slicing_configs.py (hunk 2 above)
Comment (BlackSamorez) [reply src/tensor_parallel/slicing_configs.py:103]: I've rewritten this part a bit. Now those gathers are more or less universal

File: src/tensor_parallel/slicing_configs.py [hunk 4]
```diff
@@ -78,13 +88,76 @@ def split_num_heads(num_heads: int, *, rank: int, world_size: int):
...
+                "inner_dim": partial(split_inner_dim, num_heads=model_config.num_heads, world_size=world_size),
+            },
+            r".*relative_attention_bias$": {"embedding_dim": partial(split_num_heads, world_size=world_size)},
+        },
+    )
+
+
 PREDEFINED_CONFIGS: Dict[str, ConfigGetter] = {
     "BloomModel": get_bloom_config,
+    "T5ForConditionalGeneration": get_t5_config,
```
Comment (justheuristic) [inline src/tensor_parallel/slicing_configs.py:162]: quick sanity check: are we sure this should apply to T5ForConditionalGeneration and not, say, T5Model?

**Review by justheuristic — APPROVED** [review]
Great job!

File: src/tensor_parallel/slicing_configs.py (hunk 4 above)
Comment (BlackSamorez) [reply src/tensor_parallel/slicing_configs.py:162]: There are [model architectures ](https://huggingface.co/google/flan-t5-xxl/blob/main/config.json#L3) and there is [model type](https://huggingface.co/google/flan-t5-xxl/blob/main/config.json#L15). I'm not quite sure what to use

**Pull request merged**